### PR TITLE
feat/allow users to select devices

### PR DESCRIPTION
**Pull request by juskek**
image

### CODE REVIEW

File: src/data/selected-devices/ISelectedDevicesRepository.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { DeviceName } from "../constants/DeviceEmissions";
+import { SelectedDevicesRepository } from "./SelectedDevicesRepository";
+import { TestSelectedDevicesRepository } from "./TestSelectedDevicesRepository";
+
+export abstract class ISelectedDevicesRepository {
+    private static _instance: ISelectedDevicesRepository;
```
Comment (Francisco-Costa98) [inline src/data/selected-devices/ISelectedDevicesRepository.ts:6]: underscore ? convention for private values?

File: src/data/selected-devices/SelectedDevicesRepository.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+export class SelectedDevicesRepository implements ISelectedDevicesRepository {
+    remoteDataSource: IStorageRepository = IStorageRepository.instance;
+
+    async getSelectedDevices(): Promise<Map<DeviceName, number>> {
+        try {
+            const data = await this.remoteDataSource.get({
+                selectedDevices: maptoJSON(new Map<DeviceName, number>([])),
+            });
+
+            return JSONtoMap(data["selectedDevices"] as string) as Map<
```
Comment (Francisco-Costa98) [inline src/data/selected-devices/SelectedDevicesRepository.ts:15]: not the type cast

File: src/data/page/PageRepository.ts [hunk 3]
```diff
@@ -11,8 +11,8 @@ export class PageRepository implements IPageRepository {
             });
 
             return parsePage(data["currentPage"] as string);
-        } catch (e: unknown) {
-            throw Error(e as string);
+        } catch (error: unknown) {
+            throw Error(error as string);
```
Comment (CallumHemsley) [inline src/data/page/PageRepository.ts:15]: Don't need to catch here if we will throw anyway

File: src/data/selected-countries/SelectedCountriesRepository.ts [hunk 4]
```diff
@@ -21,8 +21,8 @@ export class SelectedCountriesRepository
             return JSONtoMap(
                 data["selectedCountriesAndPercentages"] as string
             ) as Map<CountryName, number>;
-        } catch (e: unknown) {
-            throw Error(e as string);
+        } catch (error: unknown) {
+            throw Error(error as string);
```
Comment (CallumHemsley) [inline src/data/selected-countries/SelectedCountriesRepository.ts:25]: and here

File: src/data/page/PageRepository.ts [hunk 5]
```diff
@@ -21,8 +21,8 @@ export class PageRepository implements IPageRepository {
             await this.remoteDataSource.set({
                 currentPage: page,
             });
-        } catch (e: unknown) {
-            throw Error(e as string);
+        } catch (error: unknown) {
+            throw Error(error as string);
```
Comment (CallumHemsley) [inline src/data/page/PageRepository.ts:25]: and here

File: src/data/selected-countries/SelectedCountriesRepository.ts [hunk 6]
```diff
@@ -37,8 +37,8 @@ export class SelectedCountriesRepository
             await this.remoteDataSource.set({
                 selectedCountriesAndPercentages: maptoJSON(newMap),
             });
-        } catch (e: unknown) {
-            throw Error(e as string);
+        } catch (error: unknown) {
+            throw Error(error as string);
```
Comment (CallumHemsley) [inline src/data/selected-countries/SelectedCountriesRepository.ts:41]: and here

File: src/data/selected-countries/SelectedCountriesRepository.ts [hunk 7]
```diff
@@ -53,8 +53,8 @@ export class SelectedCountriesRepository
...
```
Comment (CallumHemsley) [inline src/data/selected-countries/SelectedCountriesRepository.ts:57]: and here

File: src/data/selected-countries/SelectedCountriesRepository.ts [hunk 8]
```diff
@@ -76,8 +76,8 @@ export class SelectedCountriesRepository
...
```
Comment (CallumHemsley) [inline src/data/selected-countries/SelectedCountriesRepository.ts:80]: here

File: src/data/selected-devices/SelectedDevicesRepository.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import { DeviceName } from "../constants/DeviceEmissions";
+import { IStorageRepository } from "../storage/IStorageRepository";
+import { ISelectedDevicesRepository } from "./ISelectedDevicesRepository";
+
+export class SelectedDevicesRepository implements ISelectedDevicesRepository {
+    remoteDataSource: IStorageRepository = IStorageRepository.instance;
+
+    async getSelectedDevices(): Promise<Map<DeviceName, number>> {
+        try {
+            const data = await this.remoteDataSource.get({
```
Comment (CallumHemsley) [inline src/data/selected-devices/SelectedDevicesRepository.ts:11]:
```suggestion
            const data = await this.remoteDataSource.get<Type>({
```

File: src/data/selected-devices/SelectedDevicesRepository.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+                selectedDevices: maptoJSON(new Map<DeviceName, number>([])),
+            });
+
+            return JSONtoMap(data["selectedDevices"] as string) as Map<
+                DeviceName,
+                number
+            >;
+        } catch (error: unknown) {
+            throw Error(error as string);
```
Comment (CallumHemsley) [inline src/data/selected-devices/SelectedDevicesRepository.ts:20]: throwing here

File: src/data/selected-devices/SelectedDevicesRepository.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        }
+    }
+
+    async removeSelectedDevice(device: DeviceName): Promise<void> {
+        try {
+            const newMap = await this.getSelectedDevices();
+
+            if (newMap.has(device)) {
```
Comment (CallumHemsley) [inline src/data/selected-devices/SelectedDevicesRepository.ts:44]:
```suggestion
            if (newJson[device]) {
```
don't need to have maps here potentially?

File: src/view/components/selected-devices/dropdown/useDeviceDropdown.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+    const [query, setQuery] = useState("");
+
+    const displayedDevices =
+        query === ""
+            ? filteredCountries
+            : filteredCountries.filter((country) =>
+                  country
+                      .toLowerCase()
+                      .includes(query.toLowerCase().replace(/\s+/g, ""))
```
Comment (CallumHemsley) [inline src/view/components/selected-devices/dropdown/useDeviceDropdown.ts:24]: variable or comment describing what the replace does

File: src/view/components/selected-devices/selected/useSelectedDevices.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+export const useSelectedDevices = () => {
+    const {
+        selectedDevicesContext: {
+            selectedDevices,
+            removeSelectedDevice,
+            setDevicePercentage,
+        },
+    } = useRootContext();
+    const [averagePercentage, setWorldPercentage] = useState(1);
```
Comment (CallumHemsley) [inline src/view/components/selected-devices/selected/useSelectedDevices.ts:12]: fraction?

File: src/view/provider/selected-devices/__tests__/useSelectedDevicesContext.test.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+        await act(async () => {
+            await result.current.addSelectedDevice(mockDevice2);
+        });
+
+        expect(
+            areMapsDeepEqual(result.current.selectedDevices, mockCountries)
+        ).toBe(true);
+
+        mockCountries.delete(mockDevice2);
```
Comment (CallumHemsley) [inline src/view/provider/selected-devices/__tests__/useSelectedDevicesContext.test.ts:52]: not sure why this is needed

Comment (CallumHemsley) [reply src/view/provider/selected-devices/__tests__/useSelectedDevicesContext.test.ts:52]: should be other test

File: src/view/provider/selected-devices/__tests__/useSelectedDevicesContext.test.ts [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    });
+
+    it("validatePercentages should throw an error if the sum of percentages is greater than 100%", () => {
+        const { result } = renderHook(useSelectedDevicesContext);
+
+        result.current.selectedDevices.set("iPhone 12", 0.5);
+        result.current.selectedDevices.set("Google Pixel 6", 0.6);
+
+        expect(() => result.current.validatePercentages()).toThrowError(
+            "Error: The sum of the percentages is greater than 100%. Current sum: 110%"
```
Comment (CallumHemsley) [inline src/view/provider/selected-devices/__tests__/useSelectedDevicesContext.test.ts:70]: Can this be imported from constants

File: src/view/provider/selected-devices/useSelectedDevicesContext.ts [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+                `Error: The sum of the percentages is greater than 100%. Current sum: ${(
+                    percentage * 100
+                ).toFixed(0)}%`
+            );
+        }
+    };
+
+    useMountEffect(() => {
+        const getSelectedDevicesAndSetState = async () => {
+            const _selectedDevices =
```
Comment (CallumHemsley) [inline src/view/provider/selected-devices/useSelectedDevicesContext.ts:59]: _ usually means unused variable

File: src/view/components/selected-devices/selected/useSelectedDevices.ts (hunk 13 above)
Comment (juskek) [reply src/view/components/selected-devices/selected/useSelectedDevices.ts:12]: I am intending to keep percentage in the name, so will need to refactor values to 100 - have added it to a separate ticket

File: src/view/provider/selected-devices/__tests__/useSelectedDevicesContext.test.ts (hunk 14 above)
Comment (juskek) [reply src/view/provider/selected-devices/__tests__/useSelectedDevicesContext.test.ts:52]: updated name of test for readability - I think the expect statement after adding should stay here, so the test flow is:
1. add countries
2. expect countries to be added
3. delete country
4. expect country to be deleted

File: src/data/selected-devices/SelectedDevicesRepository.ts (hunk 2 above)
Comment (juskek) [reply src/data/selected-devices/SelectedDevicesRepository.ts:15]: Added to separate ticket

**Pull request merged**